// seamlesssession2/index.js
"use strict"

this.VERSION = "1.0";
this.URI_VERSION = "version";
this.URI_OPENURI = "open-uri";
this.URI_NOTIFICATION = "notitfication";
//this.METHODS_VERSION = "get";
//this.METHODS_OPENURL = ["get", "post"];
//this.METHODS_NOTIFICATION = ["get", "post"];

var _client = class {
  constructor(api) {
    this.api = api;
    console.log("constructor: ", this.api);
  }

  version(req, res) {
    console.log("constructor: ", this.api);
    res.send({'version': this.api.VERSION});
  }

  openUri(req, res) {
    res.send({'error': 'Not yet implemented'});

    //TODO: Open the passed URI with the local default browser
  }

  notification(req, res) {
    res.send({'error': 'Not yet implemented'});

    //TODO: interface with the local notification API and cause a notification
  }

  getType() {
    return this.constructor;
  }
}

var _server = class {
  constructor(api) {
    this.api = api;
    this.clientUriBase = "https://127.0.0.1:3000/";
  }

  version(req, res) {
    res.send({'version': this.api.VERSION});
  }

  openUri(req, res) {
    //TODO: PUT: http://this.clientUriBase/api/this.api.version/this.api.URI_OPENURI?
  }

  getType() {
    return this.constructor;
  }
}

let thisClient = new _client(this);
this.apiDefination = [
  {
    'version': "1.0",
    'name': 'seamlesssession2',
    'description': 'seamlesssession2 REST API for communication between client and server agents',
    'properties': {
      'version': {
        'uri': this.URI_VERSION,
        'getter': function(req, res) {
          thisClient.version.apply(thisClient, [req, res]);
        },
        'description': 'Get the current version of this API'
      }
    },
    'methods': {
      'openuri': {
        'uri': this.URI_OPENURI,
        'action': function(req, res) {
          thisClient.openUri.apply(thisClient, [req, res]);
        },
        'description': 'Open a URI in a browser on the client\'s PC',
        'params': {
          'uri': ['string', 'The URI to open on the client\'s PC'],
          'newWindow': ['bool', 'Force the URI to be opened in a new window.  Default: false.']
        }
      },
      'notification': {
        'uri': this.URI_NOTIFICATION,
        'action': function(req, res) {
          thisClient.notification.apply(thisClient, [req, res]);
        },
        'description': 'Display a notification on the client\'s PC',
        'params': {
          'text': ['string', 'The text body of the notification'],
          'title': ['string', 'The text title of the notification.  Not supported by all client notification daemons'],
          'icon': ['string', 'The name of an icon to use'],
          'sound': ['bool', 'Play a notification sound on the client\'s PC.  Default: false.']
        }
      }
    }
  }
]

module.exports = this;
module.exports.client = _client;
module.exports.server = _server;
